From 191970aa446beb2129c7d42029e0e85af1dff876 Mon Sep 17 00:00:00 2001
From: Junxiao Chang <junxiao.chang@intel.com>
Date: Wed, 7 Dec 2022 17:03:09 +0800
Subject: [PATCH] softirq: wake up timer thread in softirq context

Timer interrupt might be triggered in softirq context, timer
thread should be waken up or else timer thread might have to wait
for next timer event which introduce timer handling latency.

Signed-off-by: Junxiao Chang <junxiao.chang@intel.com>
---
 kernel/softirq.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/kernel/softirq.c b/kernel/softirq.c
index 9bf626ebbee6..6e47354461ce 100644
--- a/kernel/softirq.c
+++ b/kernel/softirq.c
@@ -659,7 +659,7 @@ static inline void __irq_exit_rcu(void)
 	preempt_count_sub(HARDIRQ_OFFSET);
 	if (!in_interrupt() && local_softirq_pending())
 		invoke_softirq();
-	if (IS_ENABLED(CONFIG_PREEMPT_RT) && !in_interrupt() && local_pending_timers())
+	if (IS_ENABLED(CONFIG_PREEMPT_RT) && !(in_nmi() || in_hardirq()) && local_pending_timers())
 		wake_timersd();
 
 	tick_irq_exit();
-- 
2.25.1

